File: runtimes/compose/src/main/java/com/styleconverter/runtime/borders/radius/BorderRadiusApplier.kt
package com.styleconverter.runtime.borders.radius

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.RoundRect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Outline
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.LayoutDirection
import androidx.compose.ui.unit.dp
import com.styleconverter.runtime.borders.sides.AllBordersConfig
import com.styleconverter.runtime.core.types.ValueExtractors.LineStyle

/**
 * Applies CSS border-radius to a Compose modifier.
 *
 * Two code paths, chosen by [BorderRadiusConfig.isCircular]:
 *   - Circular (x == y per corner): uses the built-in [RoundedCornerShape]
 *     which is cheaper and hits Compose's fast-path corner rasterizer.
 *   - Elliptical (x != y on any corner): uses a custom [Shape] that builds
 *     a [RoundRect] with per-corner [CornerRadius] (a, b) values. This is
 *     what CSS `border-radius: 40px 20px` actually means — and why the
 *     previous (Dp-per-corner) implementation silently collapsed elliptical
 *     pairs to just the horizontal radius.
 *
 * And two APPLICATION modes, chosen by [selfPaintsWithoutClip] at the
 * StyleApplier call site (wave 42):
 *   - [applyRadius] — the legacy `Modifier.clip`, which clips EVERYTHING
 *     inner to it, child composables included. Correct only when the
 *     element actually establishes a clip (overflow != visible), and kept
 *     as the fallback for paint the self-paint mode cannot shape yet.
 *   - [applySelfPaint] — shape the element's OWN paint (background fill +
 *     uniform border band) with the corner shape and clip NOTHING.
 *     css-backgrounds-3 §4.3: "the border radii … cause the background and
 *     border to be curved" — the box's DESCENDANTS are only clipped when
 *     `overflow` says so (css-overflow-3 §3). The old always-clip path cut
 *     children away wholesale: WPT `filter-effects/backdrop-filter-
 *     clip-rect.html`'s `.navbar` (border-radius + 2px blue border) holds
 *     two red-bordered abspos `.menu` boxes that overflow it — wave-41 T5
 *     counted 1552 red pixels in the Chromium ref and in the iOS capture,
 *     and ZERO on Android (android-ref 0.8651).
 */
object BorderRadiusApplier {

    /**
     * Apply border radius by clipping the modifier to the corner shape.
     *
     * Clip is used here because this path serves the cases where the CSS
     * paint model really does bound the visible region at the rounded
     * border box: an `overflow != visible` element (css-overflow-3 §3
     * clips at the padding box, whose corners are rounded per
     * css-backgrounds-3 §4.2), and the not-yet-self-paintable combos
     * (see [selfPaintsWithoutClip]) where clipping background layers to
     * the corner shape matters more than child overflow.
     */
    fun applyRadius(modifier: Modifier, config: BorderRadiusConfig): Modifier {
        // Zero-radius fast path: no Shape allocation, no clip pass.
        if (!config.hasRadius) return modifier
        // Shared corner-shape choice (see shapeFor) wrapped in the
        // children-clipping graphicsLayer clip.
        return modifier.clip(shapeFor(config))
    }

    /**
     * The corner [Shape] for [config] — the ONE shape both application
     * modes use, so the clip path and the self-paint path can never
     * disagree about the corner geometry.
     */
    internal fun shapeFor(config: BorderRadiusConfig): Shape {
        // Circular fast path: the built-in RoundedCornerShape is the cheapest
        // and most-tested Shape; use it whenever every corner is x==y AND no
        // corner carries a paint-time percentage axis (fractions need the
        // custom Shape below to resolve against the laid-out size).
        if (config.isCircular && !config.hasFraction) {
            return RoundedCornerShape(
                topStart = config.topStart.first,
                topEnd = config.topEnd.first,
                bottomEnd = config.bottomEnd.first,
                bottomStart = config.bottomStart.first
            )
        }
        // Elliptical / percentage path: hand-built Shape that emits a
        // RoundRect with independent x/y radii per corner — this is what
        // CSS "40px 20px" and "50%" actually specify. Percentage axes are
        // resolved here against the box's width (x) / height (y) per
        // css-backgrounds-3 §4.4.
        return EllipticalCornerShape(config)
    }

    /**
     * Should this element take the non-clipping self-paint mode?
     *
     * Pure truth table (JVM-pinned in BorderRadiusSelfPaintTest). True
     * exactly when EVERY radius-shaped paint the element owns can be
     * produced by [applySelfPaint], so dropping the clip loses nothing:
     *  - a radius exists at all (otherwise neither mode installs);
     *  - the element does NOT establish an overflow clip — with
     *    `overflow: hidden/clip/scroll/auto` the children really are
     *    clipped at the rounded box (css-overflow-3 §3), which is exactly
     *    what the legacy clip renders;
     *  - no background-image layers and no background-clip insets — those
     *    paint in ColorApplier as RECTANGLES and today only look rounded
     *    because the clip cuts their corners; self-paint would un-round
     *    them. Kept on the legacy path (a deliberate, documented residual:
     *    children under gradient+radius stay clipped until the layer
     *    painter learns shapes);
     *  - the border is absent or uniform-solid — the one band
     *    [applySelfPaint] can draw exactly (Modifier.border follows the
     *    Shape). Mixed per-side widths/styles keep the legacy clip whose
     *    straight strokes + corner cut are today's committed rendering;
     *  - the element carries NO partial opacity — ColorApplier applies
     *    `opacity` as its OUTERMOST link (`Modifier.alpha`, its step 1),
     *    so the legacy chain attenuates the background it paints at its
     *    step 2, INSIDE the alpha layer. Self-paint moves the fill into
     *    THIS applier, chained OUTER of ColorApplier — under opacity < 1
     *    the rounded fill would escape the alpha layer and paint at full
     *    strength (css-color-4 §3.3 groups the element's whole paint).
     *    Kept legacy: the child cut is the smaller committed error there.
     */
    fun selfPaintsWithoutClip(
        radius: BorderRadiusConfig,
        sides: AllBordersConfig,
        overflowClips: Boolean,
        hasBackgroundLayers: Boolean,
        hasClipInsets: Boolean,
        hasPartialOpacity: Boolean,
    ): Boolean {
        // No radius → nothing for either mode to do.
        if (!radius.hasRadius) return false
        // Overflow clip requested → the children-clipping mode IS the spec.
        if (overflowClips) return false
        // Rectangular background layers would lose their rounding.
        if (hasBackgroundLayers || hasClipInsets) return false
        // Partial opacity → the fill must stay INSIDE ColorApplier's alpha
        // layer (see the truth-table bullet above); legacy owns that order.
        if (hasPartialOpacity) return false
        // Border must be self-paintable: none, or the uniform-solid band.
        return !sides.hasBorders || isUniformSolid(sides)
    }

    /**
     * The uniform-solid predicate, shared with the truth table above: all
     * four sides identical AND the DECLARED style is SOLID. `hasBorder`
     * already requires a declared visible style (BorderSideConfig.kt —
     * CSS 2.1 §8.5.3: an absent style is `none`, used width 0), so the
     * comparison is against the keyword itself; BorderSideApplier's fast
     * path and paintSide gate the same way. (retro R4 / A11#5: the old
     * note "absent style defaults to solid … like BorderSideApplier's fast
     * path" described a default neither draw path takes — the elvis was
     * dead code that misled the iOS twin into painting width-only borders
     * as solid bands.)
     */
    internal fun isUniformSolid(sides: AllBordersConfig): Boolean =
        sides.isUniform && sides.top.hasBorder && sides.top.style == LineStyle.SOLID

    /**
     * Self-paint mode: round the element's OWN paint, clip nothing.
     *
     * Two platform primitives, both Shape-aware by construction:
     *  - `Modifier.background(color, shape)` fills ONLY the shape's
     *    outline (androidx Background draws the shape's Outline path, it
     *    never clips content) — the rounded background fill of
     *    css-backgrounds-3 §4.3;
     *  - `Modifier.border(width, color, shape)` strokes the border band
     *    BETWEEN the shape's outline and its width-inset inner outline
     *    (androidx BorderModifierNode) — the curved border band of §4.3,
     *    with the inner corner curvature (outer radius minus width) the
     *    straight-stroke painter never had.
     *
     * Chain order: background FIRST (outer), border second — Compose draws
     * outer modifiers first, so the band paints on top of the fill's edge
     * exactly like the legacy chain painted strokes over the background
     * (css-backgrounds-3 §2's painting order: background, then border).
     * Children are untouched: neither primitive clips, so an overflowing
     * child paints in full (the clip-rect red menus).
     */
    fun applySelfPaint(
        modifier: Modifier,
        radius: BorderRadiusConfig,
        sides: AllBordersConfig,
        backgroundColor: Color?,
    ): Modifier {
        // One shared Shape for both paints — the same geometry the clip
        // mode would have used (see shapeFor).
        val shape = shapeFor(radius)
        var result = modifier
        // Rounded background fill, only when the element declares one
        // (ColorApplier's solid-color paint is suppressed by the caller in
        // this mode, so this is the single owner of the fill).
        if (backgroundColor != null) {
            result = result.background(backgroundColor, shape)
        }
        // Rounded uniform border band, only when the gate proved the border
        // is the uniform-solid kind this primitive renders exactly.
        if (sides.hasBorders && isUniformSolid(sides)) {
            // Same defaults as BorderSideApplier's uniform fast path
            // (width 1px, color black) so mode choice never changes the
            // resolved band.
            result = result.border(
                width = sides.top.width ?: 1.dp,
                color = sides.top.color ?: Color.Black,
                shape = shape
            )
        }
        return result
    }
}

/**
 * Custom [Shape] that produces elliptical per-corner radii by building a
 * [RoundRect] with eight independent CornerRadius components.
 *
 * Compose's [RoundedCornerShape] only accepts a single Dp per corner and
 * uses it for both axes — insufficient for CSS semantics. This Shape
 * mirrors the standard Compose corner-shape API (including RTL handling
 * via [LayoutDirection]) so it drops into `Modifier.clip` transparently.
 */
private class EllipticalCornerShape(
    private val config: BorderRadiusConfig
) : Shape {
    override fun createOutline(
        size: Size,
        layoutDirection: LayoutDirection,
        density: Density
    ): Outline {
        // Resolve logical start/end to physical left/right at paint time —
        // this matches how RoundedCornerShape handles RTL, so our custom
        // shape behaves consistently with the rest of the style engine.
        val ltr = layoutDirection == LayoutDirection.Ltr

        // Convert each corner to pixel CornerRadius(x, y). A percentage
        // axis (carried as a 0..1 fraction) resolves against the box's
        // WIDTH for x and HEIGHT for y — css-backgrounds-3 §4.4. That's
        // why `border-radius: 50%` yields a full ellipse on a non-square
        // box (rx = w/2, ry = h/2), matching Chrome. Fixed axes convert
        // Dp→px via the Density Compose hands us here.
        fun cr(
            pair: Pair<androidx.compose.ui.unit.Dp, androidx.compose.ui.unit.Dp>,
            fraction: Pair<Float?, Float?>
        ) = with(density) {
            CornerRadius(
                fraction.first?.times(size.width) ?: pair.first.toPx(),
                fraction.second?.times(size.height) ?: pair.second.toPx()
            )
        }

        // Map logical corners (start/end) to physical corners (left/right)
        // based on layout direction. In LTR: start=left, end=right.
        val cTopStart = cr(config.topStart, config.topStartFraction)
        val cTopEnd = cr(config.topEnd, config.topEndFraction)
        val cBottomEnd = cr(config.bottomEnd, config.bottomEndFraction)
        val cBottomStart = cr(config.bottomStart, config.bottomStartFraction)
        val topLeft = if (ltr) cTopStart else cTopEnd
        val topRight = if (ltr) cTopEnd else cTopStart
        val bottomRight = if (ltr) cBottomEnd else cBottomStart
        val bottomLeft = if (ltr) cBottomStart else cBottomEnd

        // RoundRect with per-corner CornerRadius(x, y) — this is the only
        // geometry type in Compose that represents true elliptical corners.
        val rr = RoundRect(
            rect = Rect(0f, 0f, size.width, size.height),
            topLeft = topLeft,
            topRight = topRight,
            bottomRight = bottomRight,
            bottomLeft = bottomLeft
        )
        // Wrap in a Path so the Outline accepts a non-convex rounded rect
        // (Outline.Rounded requires a convex RoundRect, which ours always
        // is, but Path works for any clip target and is just as fast after
        // the first rasterization).
        return Outline.Generic(Path().apply { addRoundRect(rr) })
    }
}

File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/borders/radius/BorderRadiusApplier.swift
//
//  BorderRadiusApplier.swift
//  StyleEngine/borders/radius — Phase 5, clip policy retro R4.
//
//  ViewModifier for the corner GEOMETRY's effect on the element's paint.
//  Two application modes, chosen by `clipsDescendants` (StyleBuilder
//  computes it through the static gate below):
//
//    • self-paint (default for a rounded box with overflow: visible) —
//      NO clip. The element's own paint is already curved by the
//      shape-aware appliers: ColorApplier fills BorderRadiusShape,
//      BorderSideApplier strokes it (fast paths B/C), BoxShadowApplier
//      and OutlineApplier take the radius too. css-backgrounds-3 §4.3
//      (Corner Clipping): "border images are not affected by
//      border-radius", and "other effects that clip painting or event
//      handling to the border, padding, or content edge must clip to
//      their respective curves" — i.e. the curve bounds the element's
//      OWN background (per background-clip) and any overflow clip it
//      establishes; DESCENDANTS are clipped only when `overflow` is not
//      visible (css-overflow-3 §3.1.2, Interaction of border-radius and
//      overflow) — so a child or an overflowing label paints in full
//      past the curve, as web and Compose (BorderRadiusApplier.kt
//      applySelfPaint, wave 41 T5) do. Only the hit-test region follows
//      the curve (`.contentShape`), honouring §4.3's event-handling
//      clause without cutting a pixel.
//
//    • legacy clip — `.clipShape(BorderRadiusShape)` on the whole
//      subtree, children included. Kept for exactly the cases where the
//      CSS paint model DOES bound painting at the curve, or where a paint
//      this engine still draws as a rectangle would lose its rounding
//      without the clip (see clipsDescendants).
//
//  History: Phase 5 shipped the clip UNCONDITIONALLY with a comment
//  claiming "matches CSS's overflow: hidden semantics" — which is the
//  bug: overflow: hidden is not the initial value. Every iOS-vs-other
//  pair on the visual-test rounded boxes (BorderRadius_Uniform / Pill /
//  Mixed, Edge_VeryLargeRadius, Edge_InsetRoundShadow) failed on
//  ~200 clipped label pixels with deltaE 0 — the label read "BORD" on
//  iOS and "BORDERRADIUS UNIFORM" on Android/web (retro A12#0, A11#7).
//
//  Geometry itself lives in `BorderRadiusShape`.
//

// SwiftUI for the ViewModifier API.
import SwiftUI

struct BorderRadiusApplier: ViewModifier {
    // Nil when no border-*-radius property appears in the IR.
    let config: BorderRadiusConfig?
    // TRUE → legacy `.clipShape` (children + every paint clipped to the
    // curve); FALSE → self-paint mode (no clip, hit-test shape only).
    // Defaults to TRUE so a caller that has not been taught the gate
    // renders byte-identically to Phase 5; StyleBuilder passes the
    // result of `clipsDescendants(...)` (retro R4 seam).
    var clipsDescendants: Bool = true

    func body(content: Content) -> some View {
        // Zero-radius: identity — avoids allocating a Shape for the
        // overwhelmingly common "no rounded corners" case.
        guard let cfg = config, cfg.hasAny else { return AnyView(content) }
        // ONE shape for both modes — the same BorderRadiusShape the fill
        // and stroke appliers use, so paint, clip and hit-test can never
        // disagree about the corner geometry.
        let shape = BorderRadiusShape(radius: cfg)
        if clipsDescendants {
            // Legacy / overflow-clip mode: SwiftUI `.clipShape` clips this
            // view's rendering, children included, and forwards hit-testing
            // through the shape — the css-overflow-3 §3 clip at the curved
            // padding edge (css-backgrounds-3 §4.3).
            return AnyView(content.clipShape(shape))
        }
        // Self-paint mode: paint nothing, clip nothing — the fill/stroke
        // appliers own the curved paint. `.contentShape` (SwiftUI: defines
        // the content shape for hit testing, does not affect drawing) keeps
        // §4.3's event rule without cutting a single pixel.
        return AnyView(content.contentShape(shape))
    }

    /// Should this element keep the descendant-clipping legacy mode?
    ///
    /// Pure truth table (XCTest-pinned in BorderRadiusClipPolicyTests),
    /// the iOS twin of Compose's `BorderRadiusApplier.selfPaintsWithoutClip`
    /// (inverted sense). TRUE exactly when dropping the clip would lose a
    /// paint this engine cannot yet round on its own, or when CSS says the
    /// children really are clipped:
    ///  - no radius → the modifier is identity either way (returns false);
    ///  - the element establishes an overflow clip (`overflow` ≠ visible
    ///    after css-overflow-3 §3.1 coercion — StyleBuilder.ownOverflowClips)
    ///    → css-backgrounds-3 §4.3 clips the content at the curved padding
    ///    edge, which is what the legacy clip renders;
    ///  - background-image layers paint → BackgroundImageApplier draws
    ///    gradient/url layers as RECTANGLES that today only look rounded
    ///    because the clip cuts their corners (Edge_GradientWithRadius);
    ///    the same documented residual Compose carries ("children under
    ///    gradient+radius stay clipped until the layer painter learns
    ///    shapes");
    ///  - the border does not paint along the shape by itself —
    ///    per-side / 3D / double / style-only borders take the straight-
    ///    edge Canvas path (BorderSideApplier.paintsAlongRoundedShape).
    /// Differences from the Compose table, both deliberate: no
    /// `hasClipInsets` bullet (ColorApplier's inset fill is still the
    /// BorderRadiusShape, so it keeps its rounding without the clip) and no
    /// `hasPartialOpacity` bullet (iOS applies `opacity` in
    /// applyGroupEffects, OUTER of the whole paint chain, so the fill can
    /// never escape the alpha group the way Compose's chain order allowed).
    static func clipsDescendants(radius: BorderRadiusConfig?,
                                 sides: AllBordersConfig?,
                                 overflowClips: Bool,
                                 hasBackgroundLayers: Bool) -> Bool {
        // No radius → nothing for either mode to do.
        guard let r = radius, r.hasAny else { return false }
        // Overflow clip requested → the children-clipping mode IS the spec.
        if overflowClips { return true }
        // Rectangular background layers would lose their rounding.
        if hasBackgroundLayers { return true }
        // Border must be self-rounding (absent, uniform solid, or the
        // uniform dashed/dotted rounded perimeter); otherwise keep the clip
        // whose corner cut is today's committed rendering.
        return !BorderSideApplier.paintsAlongRoundedShape(cfg: sides, radius: radius)
    }
}

// Public chain helper, following the `.engine*` naming convention.
extension View {
    // `clipsDescendants` defaults to the legacy clip so the un-patched
    // call shape keeps compiling and rendering unchanged; StyleBuilder
    // passes `BorderRadiusApplier.clipsDescendants(...)` (retro R4).
    func engineBorderRadius(_ config: BorderRadiusConfig?,
                            clipsDescendants: Bool = true) -> some View {
        modifier(BorderRadiusApplier(config: config, clipsDescendants: clipsDescendants))
    }
}
